### PR TITLE
Some cleanup in bytecomp/emitcode.ml

### PR DESCRIPTION
**Pull request by hhugo**
The PR tries to clarify/fix some logic in emitcode.ml.
It's not really a bugfix because of implicit invariants in the code. 

- Both `init ()` and `reset ()` were doing a reset in a slightly different way, now merged.
- `extend_label_table` only worked when the initial table is of size > 0. (note that `reset` was previously setting an empty table, later overwritten by `init`)
- the logic around `out_buffer` only worked when original buffer size was >= 4. ( a size < 1 would make the buffer growing logic loop forever. A size < 4 would make `out_word` write out of bound)

### CODE REVIEW

File: bytecomp/emitcode.ml [hunk 1]
```diff
@@ -54,14 +54,17 @@ let out_buffer_size = 1024
 let out_buffer = ref(LongString.create out_buffer_size)
 and out_position = ref 0
 
+let extend_bufffer needed =
```
Comment (lthls) [inline bytecomp/emitcode.ml]: I think you might have extended the array of `f`s by mistake

**Review by lthls — COMMENTED** [review]
All of the functions exposed in the interface of this file start by calling `init` (or `reset` in your PR), and call it again before they return; wouldn't it make sense to simply remove `reset` from the interface ?

I also think the `extend_buffer` function might be slightly overkill (unlike the label table, we only ever grow `out_buffer` by small increments, so we probably don't need very complicated logic). But the while loop is cheap compared to the rest of the code in this branch, so I think it's fine.

**Comment by hhugo** [comment]
> I also think the extend_buffer function might be slightly overkill.

The while loop is necessary if you start from a buffer of size 1. Such change makes the buffer code not depend on implicit invariant present in this file.

File: bytecomp/emitcode.ml (hunk 1 above)
Comment (hhugo) [reply bytecomp/emitcode.ml]: fixed

**Comment by hhugo** [comment]
> wouldn't it make sense to simply remove reset from the interface ?

I've now removed `reset` from the interface.

**Review by lthls — APPROVED** [review]
Looks fine to me. I think you should add a Changes entry though.

**Comment by hhugo** [comment]
I've pushed an additional change to clear the global state on exception.

Will add an changelog entry after you approve this last change.

File: bytecomp/emitcode.ml [hunk 2]
```diff
@@ -446,21 +446,21 @@ let to_file outchan unit_name objfile ~required_globals code =
...
 
 let to_packed_file outchan code =
   init ();
+  Fun.protect ~finally:clear (fun () ->
   emit code;
   LongString.output outchan !out_buffer 0 !out_position;
   let reloc = !reloc_info in
-  clear ();
-  reloc
+  reloc)
```
Comment (lthls) [inline bytecomp/emitcode.ml:469]: You could remove the last let-binding if you want (otherwise the compiler will do it for you anyway).

**Review by lthls — APPROVED** [review]
The patch looks correct. Could you give a reason why you didn't protect `to_file` ? Is it only because the patch would have been less trivial ?

**Comment by gasche** [comment]
In your changelog, can you also include #11601 maybe?

**Comment by hhugo** [comment]
> The patch looks correct. Could you give a reason why you didn't protect `to_file` ? Is it only because the patch would have been less trivial ?

Because I forgot to get back to it. It was not clear to me what should be done with [Btype.cleanup_abbrev ()]. I've updated the comment about it to make it clearer to future reviewers.

`to_file` is now protected as well.

File: bytecomp/emitcode.ml (hunk 2 above)
Comment (hhugo) [reply bytecomp/emitcode.ml:469]: I've some other PR coming for this, if you don't mind to wait

**Comment by hhugo** [comment]
> In your changelog, can you also include #11601 maybe?

I've added a single entry for both, giving room to other PRs. cc @smuenzel

**Comment by hhugo** [comment]
@lthls, @gasche are we good to go ?

**Comment by smuenzel** [comment]
I think Changes has merge conflicts?

**Comment by hhugo** [comment]
I've fixed the conflict. You should probably squash all commits

**Comment by lthls** [comment]
Thanks, and sorry for the delay.

**Pull request merged**